push_notifications/services/apns.py

```diff
@@ -0,0 +1,239 @@
+"""
+Apple Push Notification Service
+Documentation is available on the iOS Developer Library:
+https://developer.apple.com/library/ios/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/ApplePushService.html
+"""
+
+import codecs
+import json
+import ssl
+import struct
+import socket
+import time
+from contextlib import closing
+from binascii import unhexlify
+from django.core.exceptions import ImproperlyConfigured
+from .base import BaseService
+from .exceptions import NotificationError
+
+
+class APNSError(NotificationError):
+    pass
+
+
+class APNSServerError(APNSError):
+    def __init__(self, status, identifier):
+        super(APNSServerError, self).__init__(status, identifier)
+        self.status = status
+        self.identifier = identifier
+
+
+class APNSDataOverflow(APNSError):
+    pass
+
+
+class APNSPushService(BaseService):
+
+    def __init__(self, configuration):
+        self.settings = configuration.get('APNS_CONF', {})
+
+        self.settings.setdefault('APNS_PORT', 2195)
+        self.settings.setdefault('APNS_FEEDBACK_PORT', 2196)
+        self.settings.setdefault('APNS_ERROR_TIMEOUT', None)
+        self.settings.setdefault('APNS_MAX_NOTIFICATION_SIZE', 2048)
+        if self.settings.get('APNS_DEBUG', False):
+            self.settings.setdefault('APNS_HOST', 'gateway.sandbox.push.apple.com')
+            self.settings.setdefault('APNS_FEEDBACK_HOST', 'feedback.sandbox.push.apple.com')
+        else:
+            self.settings.setdefault('APNS_HOST', 'gateway.push.apple.com')
+            self.settings.setdefault('APNS_FEEDBACK_HOST', 'feedback.push.apple.com')
+
+        if not self.settings.get('APNS_CERTIFICATE'):
+            raise ImproperlyConfigured('For APNS service to work you need an '
+                                       'APNS_CERTIFICATE in the configuration')
+
+    def _create_socket(self, address_tuple):
+        certfile = self.settings.get("APNS_CERTIFICATE")
+        if not certfile:
+            raise ImproperlyConfigured(
+                'You need to set DJANGO_PUSH_NOTIFICATIONS["APNS_CONF"]["APNS_CERTIFICATE"]\
+                 to send messages through APNS.'
+            )
+
+        try:
+            with open(certfile, "r") as f:
+                f.read()
+        except Exception as e:
+            raise ImproperlyConfigured("The APNS certificate file at %r is not readable: %s" % (certfile, e))
+
+        ca_certs = self.settings.get("APNS_CA_CERTIFICATES")
+
+        sock = socket.socket()
+        sock = ssl.wrap_socket(sock, ssl_version=ssl.PROTOCOL_TLSv1, certfile=certfile, ca_certs=ca_certs)
+        sock.connect(address_tuple)
+
+        return sock
+
+    def _create_socket_to_push(self):
```

Comment:
I would strongly insist on using a third-party library like pyAPNS for this, instead of maintaining low level logic in this project.